### PR TITLE
Alt Click Madness

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Adds a number of alt click behaviors to some commonly used verbs (that are otherwise annoying to use).

- Adds AltClick rotating of:
 - Chairs
 - Disposal pipes
 - Windows
 - The PA
 - Emitters
 - Windoor Assemblies

Now you don't have to right click or rely on your crummy objects verb for some of these tasks

:cl: Fox McCloud
add: adds alt-click rotating of chairs, disposal pipes, windows, the PA, emitters, and windoor assemblies
/:cl:

### CODE REVIEW

**Comment by Spacemanspark** [comment]
:+1:

**Comment by EpicCharger** [comment]
You mean I don't have to right click the normal window set up, find out which window I'm trying to rotate, and then rotate it? Praise be. Will this have conflicts with the alt click to show everything on that tile? Say, if there was a disposal pipe and things under it, and I wanted to get to the things under it by using the alt click to view all on the tile, would the rotating cancel it out?

**Comment by Fox-McCloud** [comment]
It will show everything on the tile AND rotate it.

**Comment by EpicCharger** [comment]
Lovely. Just gotta be careful rotating an item sitting on a thousand other items! Sounds like a PR that'll save me some annoyance.

File: code/game/objects/structures/stool_bed_chair_nest/chairs.dm [hunk 1]
```diff
@@ -66,6 +66,16 @@
...
 		return
 
+/obj/structure/stool/bed/chair/AltClick(mob/user)
+	..()
+	if(user.incapacitated())
+		to_chat(user, "<span class='warning'>You can't do that right now!</span>")
+		return
+	if(!in_range(src, user))
+		return
+	else
```
Comment (PJB3005) [inline code/game/objects/structures/stool_bed_chair_nest/chairs.dm]: No need for this else clause.

File: code/game/objects/structures/windoor_assembly.dm [hunk 2]
```diff
@@ -322,6 +322,16 @@ obj/structure/windoor_assembly/Destroy()
...
 	return
 
+/obj/structure/windoor_assembly/AltClick(mob/user)
+	..()
+	if(user.incapacitated())
+		user << "<span class='warning'>You can't do that right now!</span>"
+		return
+	if(!in_range(src, user))
+		return
+	else
```
Comment (PJB3005) [inline code/game/objects/structures/windoor_assembly.dm]: Ditto.

File: code/game/objects/structures/windoor_assembly.dm [hunk 3]
```diff
@@ -322,6 +322,16 @@ obj/structure/windoor_assembly/Destroy()
 	update_icon()
 	return
 
+/obj/structure/windoor_assembly/AltClick(mob/user)
+	..()
+	if(user.incapacitated())
+		user << "<span class='warning'>You can't do that right now!</span>"
```
Comment (PJB3005) [inline code/game/objects/structures/windoor_assembly.dm]: `to_chat`

File: code/game/objects/structures/window.dm [hunk 4]
```diff
@@ -318,6 +318,16 @@ var/global/wcCommon = pick(list("#379963", "#0d8395", "#58b5c3", "#49e46e", "#8f
...
 
 
+/obj/structure/window/AltClick(mob/user)
+	..()
+	if(user.incapacitated())
+		to_chat(user, "<span class='warning'>You can't do that right now!</span>")
+		return
+	if(!in_range(src, user))
+		return
+	else
```
Comment (PJB3005) [inline code/game/objects/structures/window.dm]: More else clause madness.

File: code/modules/power/singularity/emitter.dm [hunk 5]
```diff
@@ -70,6 +70,16 @@
...
 	return 1
 
+/obj/machinery/power/emitter/AltClick(mob/user)
+	..()
+	if(user.incapacitated())
+		to_chat(user, "<span class='warning'>You can't do that right now!</span>")
+		return
+	if(!in_range(src, user))
+		return
+	else
```
Comment (PJB3005) [inline code/modules/power/singularity/emitter.dm]: Ahem.

File: code/modules/power/singularity/particle_accelerator/particle_accelerator.dm [hunk 6]
```diff
@@ -101,6 +101,16 @@ So, hopefully this is helpful if any more icons are to be added/changed/wonderin
...
 	return 1
 
+/obj/structure/particle_accelerator/AltClick(mob/user)
+	..()
+	if(user.incapacitated())
+		to_chat(user, "<span class='warning'>You can't do that right now!</span>")
+		return
+	if(!in_range(src, user))
+		return
+	else
```
Comment (PJB3005) [inline code/modules/power/singularity/particle_accelerator/particle_accelerator.dm]: :tada:

File: code/modules/recycling/disposal-construction.dm [hunk 7]
```diff
@@ -96,6 +96,16 @@
...
 	update()
 
+/obj/structure/disposalconstruct/AltClick(mob/user)
+	..()
+	if(user.incapacitated())
+		to_chat(user << "<span class='warning'>You can't do that right now!</span>")
+		return
+	if(!in_range(src, user))
+		return
+	else
```
Comment (PJB3005) [inline code/modules/recycling/disposal-construction.dm]: Did I mention else clauses?

File: code/game/objects/structures/stool_bed_chair_nest/chairs.dm [hunk 8]
```diff
@@ -66,6 +66,16 @@
 		handle_rotation()
 		return
 
+/obj/structure/stool/bed/chair/AltClick(mob/user)
+	..()
+	if(user.incapacitated())
+		to_chat(user, "<span class='warning'>You can't do that right now!</span>")
+		return
+	if(!in_range(src, user))
```
Comment (PJB3005) [inline code/game/objects/structures/stool_bed_chair_nest/chairs.dm]: Should check adjacency instead.

File: code/game/objects/structures/windoor_assembly.dm [hunk 9]
```diff
@@ -322,6 +322,16 @@ obj/structure/windoor_assembly/Destroy()
...
+		return
+	if(!in_range(src, user))
```
Comment (PJB3005) [inline code/game/objects/structures/windoor_assembly.dm]: Should check adjacency instead.

File: code/game/objects/structures/window.dm [hunk 10]
```diff
@@ -318,6 +318,16 @@ var/global/wcCommon = pick(list("#379963", "#0d8395", "#58b5c3", "#49e46e", "#8f
 	return
 
 
+/obj/structure/window/AltClick(mob/user)
+	..()
+	if(user.incapacitated())
+		to_chat(user, "<span class='warning'>You can't do that right now!</span>")
+		return
+	if(!in_range(src, user))
```
Comment (PJB3005) [inline code/game/objects/structures/window.dm]: Should check adjacency instead.

File: code/modules/power/singularity/emitter.dm [hunk 11]
```diff
@@ -70,6 +70,16 @@
 	src.dir = turn(src.dir, 90)
 	return 1
 
+/obj/machinery/power/emitter/AltClick(mob/user)
+	..()
+	if(user.incapacitated())
+		to_chat(user, "<span class='warning'>You can't do that right now!</span>")
+		return
+	if(!in_range(src, user))
```
Comment (PJB3005) [inline code/modules/power/singularity/emitter.dm]: Should check adjacency instead.

File: code/modules/power/singularity/particle_accelerator/particle_accelerator.dm [hunk 12]
```diff
@@ -101,6 +101,16 @@ So, hopefully this is helpful if any more icons are to be added/changed/wonderin
 	src.dir = turn(src.dir, 270)
 	return 1
 
+/obj/structure/particle_accelerator/AltClick(mob/user)
+	..()
+	if(user.incapacitated())
+		to_chat(user, "<span class='warning'>You can't do that right now!</span>")
+		return
+	if(!in_range(src, user))
```
Comment (PJB3005) [inline code/modules/power/singularity/particle_accelerator/particle_accelerator.dm]: Should check adjacency instead.

File: code/modules/recycling/disposal-construction.dm [hunk 13]
```diff
@@ -96,6 +96,16 @@
 	dir = turn(dir, -90)
 	update()
 
+/obj/structure/disposalconstruct/AltClick(mob/user)
+	..()
+	if(user.incapacitated())
+		to_chat(user << "<span class='warning'>You can't do that right now!</span>")
+		return
+	if(!in_range(src, user))
```
Comment (PJB3005) [inline code/modules/recycling/disposal-construction.dm]: Should check adjacency instead.

**Comment by PJB3005** [comment]
Little nitpick: I'd suggest you space your code out more, especially after the end of a code block.

**Comment by Davdichan** [comment]
What about alt clicking PDAs to remove IDs like /tg/ has? Also alt clicking lockers with an empty hand to lock them quickly and easily.

**Comment by monster860** [comment]
@Davdichan Lockers can be locked easily just by holding an item.

**Comment by Davdichan** [comment]
@monster860 Not really a valid excuse, person shouldn't need to smack a locker with their PDA or oxygen tank to lock it when a simple key press combo would let them lock them quicker and without having to empty their pockets.

**Comment by Taaketa** [comment]
I will have your babies. Even though I am biologically incapable of it!

**Comment by theColdflame** [comment]
@Davdichan you can remove IDs from PDAs with altclick on paracode as of the previous patch

**Comment by Lady-Luck** [comment]
Yano, those firelock popups are annoying as shit, could that be moved to alt click functionality as well?

**Comment by FlattestGuitar** [comment]
Hardsuit helmet toggle pls. 

Thank you.

**Comment by Fox-McCloud** [comment]
>Hardsuit helmet toggle pls.

Thank you.

Hell no. That belongs in the realm of an action button.

Eventually I'll kill off the current implementation of hardsuits in favor of TG's (no attaching helmets, action buttons for toggling+sounds, built in jetpacks, etc), but it requires a decent amount of groundwork to be laid, first.

**Comment by FlattestGuitar** [comment]
I am absolutely fine with that.

**Comment by Plotron** [comment]
We need dem damned firelocks relegated to alt-click, please! It's... so... fucking... annoying!
I also agree about the lockers.

What about closing fire extinguisher wall-mounted closets? Could also be alt-click.

**Comment by Krausus** [comment]
> It will show everything on the tile AND rotate it.

Oh, geez, don't add new clicks that work like this. It means trying to alt-click something in the world to manipulate it will hijack your current stat tab, and trying to alt-click something in the "turf contents" tab will close it.

Alt-click is fine for stuff in the interface, such as the PDA ID removal, but stuff in-world should use a click combo that's not already used. Ctrl-click would probably work fine, or, since you're replacing verbs you'd normally right-click for, you could even do something like alt-right-click.

But please, don't put alt-click on double duty.

**Comment by Fox-McCloud** [comment]
The problem with that is, `control` is already used up by something that would be far more disruptive, since it makes you start pulling things. Likewise, shift click makes you examine things.

Yes, alt-click is pulling double duty, but it's far less disruptive, overall, to performing such a task than than the other two are----and control/alt/shift+alt/shift/control clicking is highly awkward, especially for those who don't use non-hotkey mode---and it's far less intrusive then right clicking, waiting for something to pop up, then menu baring over and finally selecting the actual verb (even more so if you have to click something more than once, as per windows).

**Comment by Krausus** [comment]
Oh, right, most of this stuff can be pulled. Derp.

Alt+right click should be fine, though. Just make it perform the action without popping up the menu. Alt-right-clicking doesn't even properly pop up the menu as it is - it pops up and immediately disappears.

**Comment by Davdichan** [comment]
Doesn't the tg code check what you are clicking on alt tab? If you click a non-rotable/manipulative object while holding alt, it opens the menu, but if you are hovering over something that can it doesn't open the menu and just does the option on that object. I know you can alt rotate objects in the examine window without closing that too.

**Comment by Krausus** [comment]
Welp... turns out BYOND doesn't actually let you process right-clicks unless you disable the context menu entirely, which *may* be feasible with some other changes, but is definitely outside the scope of this PR.

Still, making alt-click perform some function and *also* pop up the turf contents isn't ideal. I suggest this: have alt-clicking an alt-clickable not call its parent unless you're returning with zero feedback, so turf contents don't pop up unless literally nothing happened; and, make shift+alt+click *always* pop up the turf contents, no matter what you click on.

The only thing that uses shift+alt+click currently is toggling emergency override on doors, and that would only matter if a cyborg happens to want to see the contents of a turf with a closed door obscuring everything on it.

**Comment by Fox-McCloud** [comment]
Removed the call to parent.

If this causes it to function in a frustrating manner, don't blame me for it.

**Pull request merged**